internal/transport/http2_client.go

```diff
@@ -64,6 +64,7 @@ type http2Client struct {
 	cancel     context.CancelFunc
 	ctxDone    <-chan struct{} // Cache the ctx.Done() chan.
 	userAgent  string
+	address    resolver.Address // Record the used resolver address of client, and replace :authority to resolver address if serverName is not empty.
```

Comment:
Nit: end-of-line comments are usually short. So, maybe something like this would suffice here (with a more detailed comment in `NewStream` where the value is used):

```golang
  address resolver.Address // Used to set :authority header
  ```
  
  Or a proper comment sentence might also be used:
```golang
  // address contains the resolver returned address for this transport.
  // If the `ServerName` field is set, it takes precedence over `CallHdr.Host`
  // passed to `NewStream`, when determining the :authority header.
  address resolver.Address 
```